### DISCUSSION TITLE
Firefly started forgetting access tokens

### DISCUSSION THREAD

**Discussion by avee87** [firefly-iii/firefly-iii#9396] [Firefly III]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem
- [x] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

On Monday I noticed that my Firefly instance "forgot" all personal access tokens. They still show up in the list but applications such as Importer can no longer authenticate:
> Error accessing "https://firefly.my-domain.com/api/v1/about?". Status code is 401. Body is: {"message":"Unauthenticated","exception":"AuthenticationException"}

I generated new tokens which seemed to fix the issue but today they stopped working again. I suspect it may coincide with instance restarts but need to test it more.

Did something change recently wrt how Firefly stores these tokens?

### Debug information

Debug information generated at 2024-10-23 12:23:31 for Firefly III version **v6.1.21**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>6.1.21 / v2.1.0 / 24 (exp. 24)</td></tr><tr><td>PHP version</td><td>8.3.12 (64bits) / apache2handler / Linux x86_64</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td><strong>*mysql*</strong>, pgsql, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>1131, base <span>#</span>91</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>Europe/London + Europe/London</td></tr><tr><td>App environment</td><td>local, debug: false</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>notice, stack / (empty)</td></tr><tr><td>Cache driver</td><td>file</td></tr><tr><td>Default language and locale</td><td>en_GB + equal</td></tr><tr><td>Trusted proxies</td><td>**</td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>2024-10-22 23:00:00 (12 hours ago)</td></tr><tr><td>Mailer</td><td>smtp</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 1</td></tr><tr><td>User flags</td><td><span title="Has liabilities">:credit_card:</span> <span title="Has piggy banks">:pig:</span> <span title="Has rules">:wrench:</span></td></tr><tr><td>Session start</td><td>2024-10-01 00:00:00</td></tr><tr><td>Session end</td><td>2024-10-31 23:59:59</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>en_GB</td></tr><tr><td>User locale</td><td>en_GB</td></tr><tr><td>Locale(s) supported</td><td>en_GB.utf8: :white_check_mark:<br>en_GB.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36</td></tr></tbody></table>

**Comment by JC5** [comment]
Nothing has changed for the longest time, and my own access tokens hold up fine (so far). They do expire after about one year, that could be it?

**Comment by avee87** [comment]
Hm, this is interesting... One of the tokens was about a year old but another one was much younger - yet they both stopped working. And both "died" again just 2 days later. Will try rotating them again and report if the issue repeats.

**Comment by JC5** [comment]
Let me know. You can certainly "kill" access tokens if the public/private keypair is destroyed. This is something that [Monica](https://github.com/monicahq/monica) suffers from for example, each container reboot kills all access tokens.

However, Firefly III stores these keys in your configuration table, so they are restored after a reboot. If this process fails it could mean the access tokens get destroyed. However, this is pretty foolproof as far as I've been able to check.

Shall I convert this into a discussion in the mean time?

**Comment by quetzai** [comment]
I had the same thing happening to me on the same version of FireflyIII.
Rotating keys worked, once I'm done using it I'll report on the behaviour after:
- `docker compose restart` (I expect no problem there)
- `docker compose up` (might trigger something?

**Reply by JC5** [reply]
`docker compose stop`, `start`, `restart` will not give problems, I use those all the time.

`docker compose up` is paired with `docker compose down`, and that command stops containers and removes containers, networks, volumes, and images created by `up`.

So yeah when all your data is deleted you will need new keys.

**Reply by iLuckyW** [reply]
Hey, little late to the discussion. I recently noticed the same issue. Is there an option to mount a path in docker to a local folder, so that the person access tokens are persistent after container deletion?

**Reply by JC5** [reply]
Access tokens are already persisted across container deletion. What are you running into?

**Reply by iLuckyW** [reply]
I am sorry, yeah they are also persistent for me, now that I tried this specifically. I just had a few token become invalid in the last months, just like @avee87 describe in his post. Thought it had to do with deleting and recreating the container, as you mentioned in a post above.
Unfortunately I was not able to exactly determine when and why the token got invalid. Maybe I can spot the next time it happens
